src/Compilers/Helpers/CoreClrAnalyzerAssemblyLoader.cs

```diff
@@ -0,0 +1,163 @@
+            }
+        }
+
+        public Assembly LoadFromPath(string fullPath)
+        {
+            if (fullPath == null)
+            {
+                throw new ArgumentNullException(nameof(fullPath));
+            }
+
+            Debug.Assert(PathUtilities.IsAbsolute(fullPath));
+
+            lock (_guard)
+            {
+                Assembly assembly;
+                if (_pathsToAssemblies.TryGetValue(fullPath, out assembly))
+                {
+                    return assembly;
+                }
+
+                return LoadAndCache(fullPath);
+            }
+        }
+
+        private static readonly string[] _extensions = new string[] { ".dll", ".exe" };
+
+        /// <summary>
+        /// Searches and loads from the base directory of the current
+        /// app context
+        /// </summary>
+        private Assembly AppContextLoad(AssemblyName assemblyName)
+        {
+            var baseDir = AppContext.BaseDirectory;
+            foreach (var extension in _extensions)
+            {
+                var path = Path.Combine(baseDir, assemblyName.Name + extension);
+
+                if (File.Exists(path))
+                {
+                    return LoadAndCache(path);
+                }
+            }
+            return null;
+        }
+
+        protected override Assembly Load(AssemblyName assemblyName)
+        {
+            lock (_guard)
+            {
+                // Try and grab assembly using standard load
+                Assembly assembly = AppContextLoad(assemblyName);
+                if (assembly != null)
+                {
+                    return assembly;
+                }
+
+                string fullName = assemblyName.FullName;
+
+                if (_namesToAssemblies.TryGetValue(fullName, out assembly))
+                {
+                    return assembly;
+                }
+
+                AssemblyIdentity requestedIdentity;
+                if (!AssemblyIdentity.TryParseDisplayName(fullName, out requestedIdentity))
+                {
+                    return null;
+                }
+
+                foreach (var candidatePath in _dependencyPaths)
+                {
+                    if (AssemblyAlreadyLoaded(candidatePath) ||
+                        !FileMatchesAssemblyName(candidatePath, requestedIdentity.Name))
+                    {
+                        continue;
+                    }
+
+                    var candidateIdentity = TryGetAssemblyIdentity(candidatePath);
+
+                    if (requestedIdentity.Equals(candidateIdentity))
+                    {
+                        return LoadAndCache(candidatePath);
+                    }
+                }
+
+                return null;
+            }
+        }
+
+        /// <remarks>
+        /// Assumes we have a lock on _guard
+        /// </remarks>
+        private Assembly LoadAndCache(string fullPath)
+        {
+            var assembly = LoadFromAssemblyPath(fullPath);
```

Comment:
Can `LoadFromAssemblyPath` return `null`?  Several of our impl methods which I believe are called as a part of this do return `null`.